lmdeploy/serve/qos_engine/qos_engine.py

```diff
@@ -0,0 +1,241 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import asyncio
+import json
+import logging
+import threading
+import time
+
+from lmdeploy.serve.openai.protocol import (ChatCompletionRequestQos,
+                                            CompletionRequestQos,
+                                            GenerateRequestQos)
+from lmdeploy.serve.qos_engine.inner_group_schd import UserRequestQueue
+from lmdeploy.serve.qos_engine.usage_stats import UsageStats
+
+logger = logging.getLogger(__name__)
+
+
+class QosConfig:
+
+    def __init__(self, qos_tag=''):
+        try:
+            qos_config = json.loads(qos_tag)
+            self.is_qos_enabled = qos_config['enable_user_qos']
+            self.user_id_maps = qos_config['user_group_map']
+            self.user_group_prio = qos_config['user_groups']
+        except Exception:
+            self.is_qos_enabled = False
+            self.user_id_maps = dict()
+            self.user_group_prio = []
+        logger.debug(f'is_qos_enabled: {self.is_qos_enabled}')
+        logger.debug(f'user_id_maps:  {self.user_id_maps}')
+        logger.debug(f'user_group_prio: {self.user_group_prio}')
+
+
+class QosEngine:
+
+    def __init__(self, qos_tag='', engine=None, **kwargs) -> None:
+        self.engine = engine
+        self.availSlots = engine.instance_num
+        self._stop_event = threading.Event()
+        self._dequeue_thread = threading.Thread(target=self._serve,
+                                                daemon=True)
+        self.qos_config = QosConfig(qos_tag)
+
+        self.qos_user_group = QosGroupQueue(self.qos_config)
+
+        self.usage_stats = UsageStats(60, 6, 0,
```

Comment:
May use `key=value` to initiate `UsageStats` in order to clarify what 60, 6, 0 are

Reply:
okk